File: Python/week_6/sentimental-credit/credit.py
# TODO
from cs50 import get_int


def main():

    # Prompt user for input
    card = user_prompt()

    # Check decimal places
    decimal_places = how_many_decimal_places(card)

    # Checksum algorithm
    validity = checksum(card, decimal_places)

    # Check card length and digits
    card_type = check_card_type(card, decimal_places)

    # Print AMEX, VISA, MASTERCARD or INVALID according to checksum and card type
    validation_print(card_type, validity)


def user_prompt():
    # get user card number
    userInput = get_int("Number: ")
    return userInput


def checksum(card, decimal):

    # sum of even numbers for checksum
    suma_pares = 0
    # sum of odd numbers for cheksum
    suma_impares = 0
    # prepare variable to run for loop according to length of card
    length = decimal - 1
    counter = 1

    # iterate over card numbers
    for i in range(length, -1, -1):
        # par
        temporary = str(card)

        # discern if even or odd
        if (counter % 2 == 0):
            temporary_even_variable = int(temporary[i])

            # if even, check if multiplicaton of number by 2 creates 2 digits ('10' or more)
            if (temporary_even_variable * 2 > 9):

                # if 2 digits, treate each digit as a separate number
                temporary_even_variable *= 2
                first_digit = str(temporary_even_variable)[0]
                second_digit = str(temporary_even_variable)[1]

                # add separate numbers to sum of even numbers for checksum
                suma_pares = int(first_digit) + int(second_digit) + suma_pares

            else:
                # if not add to sum of pairs for checksum
                temporary_even_variable *= 2
                suma_pares = temporary_even_variable + suma_pares

        # impar
        else:
            # if odd, add to sum of odd numbers for checksum
            temporary_odd_variable = int(temporary[i])
            suma_impares += temporary_odd_variable

        counter += 1

    # add even and odd numbers to get checksum
    checksum = suma_pares + suma_impares

    # return if last digit of checksum is a 0
    if (checksum % 10 == 0):
        return 1
    else:
        return 0


def check_card_type(card, decimal_places):

    # VISA
    if (decimal_places == 13):

        # get first digit of card
        first_digit = str(card)[0]
        first_digit = int(first_digit)

        # check if first digit of card is 4
        if (first_digit == 4):
            return "VISA"

        else:
            return "INVALID"

    # AMEX
    elif (decimal_places == 15):

        first_digit = str(card)[0]
        second_digit = str(card)[1]
        check_digits = int(first_digit + second_digit)

        # check if first numbers of card are 34 or 37
        if (check_digits == 34 or check_digits == 37):
            return "AMEX"
        else:
            return "INVALID"

    # MASTERCARD OR VISA
    elif (decimal_places == 16):

        check_digits = card / 10 ** 14

        # check if first numbers of card are numbers 50 - 56 for mastercard
        if (check_digits > 50 and check_digits < 56):
            return "MASTERCARD"

        # check if first numbers of card are numbers 39 - 50 for visa
        elif (check_digits > 39 and check_digits < 50):
            return "VISA"

        else:
            return "INVALID"

    # INVALID
    else:
        return "INVALID"


def validation_print(card_type, validity):

    # print if valid, the card's type, else invalid
    if (validity == 1):
        print(f"{card_type}")

    else:
        print("INVALID")


def how_many_decimal_places(number):
    decimal_places = len(str(number))

    return decimal_places


main()

File: week_1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

long user_prompt(void);
int checksum(long card, int decimal);
string check_card_type(long card, int decimal_places);
void validation_print(string card_type, int validity);
int how_many_decimal_places(long n);



int main(void)
{
    //Prompt user for input (get_long)
    long card = user_prompt();

    //Check decimal places
    int decimal_places = how_many_decimal_places(card);

    //Checksum Algorithm
    int validity = checksum(card, decimal_places);

    //Check Card length and digits
    string card_type = check_card_type(card, decimal_places);

    //Print AMEX, VISA, MASTERCARD or INVALID
    validation_print(card_type, validity);
}




















long user_prompt(void)
{
    long n = get_long("Number: ");
    return n;
}


int checksum(long card, int decimal)
{
    int checksum, suma_pares = 0, suma_impares = 0, o, par;
    int w, x, y, z;
    long a, b, m;

    for (int i = 0; i < decimal ; i++)
    {
        a = pow(10, i);
        b = pow(10, i + 1);

        m = card % a;

        o = ((card % b) - m) / a;

        //Pares
        if ((i + 1) % 2 == 0)
        {
            par = o * 2;

            if (par > 9)
            {
                for (int j = 0; j < 2 ; j++)
                {
                    x = pow(10, j);
                    y = pow(10, j + 1);

                    w = par % x;

                    z = ((par % y) - w) / x;

                    suma_pares += z;
                }
            }
            else
            {
                suma_pares += par;
            }
        }
        //Impares
        else
        {
            suma_impares += o;
        }
    }

    //Suma de pares e impares
    checksum = suma_pares + suma_impares;


    //Checksum
    if (checksum % 10 == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}


string check_card_type(long card, int decimal_places)
{
    string invalid = "INVALID", visa = "VISA", mastercard = "MASTERCARD", american_express = "AMEX";

    //VISA
    if (decimal_places == 13)
    {
        int check_digits = card / pow(10, 12);
        if (check_digits == 4)
        {
            return visa;
        }
        else
        {
            return invalid;
        }
    }
    //AMEX
    else if (decimal_places == 15)
    {
        int check_digits = card / pow(10, 13);
        if (check_digits == 34 || check_digits == 37)
        {
            return american_express;
        }
        else
        {
            return invalid;
        }
    }
    //MASTERCARD OR VISA
    else if (decimal_places == 16)
    {
        int check_digits = card / pow(10, 14);
        if (check_digits > 50 && check_digits < 56)
        {
            return mastercard;
        }
        else if (check_digits > 39 && check_digits < 50)
        {
            return visa;
        }
        else
        {
            return invalid;
        }
    }
    //INVALID
    else
    {
        return invalid;
    }
}


void validation_print(string card_type, int validity)
{
    if (validity == 1)
    {
        printf("%s\n", card_type);
    }
    else
    {
        printf("INVALID\n");
    }
}


int how_many_decimal_places(long n)
{
    int counter = 0;
    float check = 0;
    long power_of_ten = pow(10, counter);

    do
    {
        check = n / power_of_ten;
        counter++;
        power_of_ten = pow(10, counter);
    }
    while (check > 1);

    counter--;

    return counter;
}